packages/core/src/traceCollection.ts

```diff
@@ -0,0 +1,143 @@
+import {
+  Batch,
+  Configuration,
+  Context,
+  getTimestamp,
+  HttpRequest,
+  msToNs,
+  RequestCompleteEvent,
+  startRequestCollection,
+} from '.'
+import { UserConfiguration } from './configuration'
+
+export interface SpanMetadata {
+  [key: string]: string
+}
+
+export interface SpanMetrics {
+  [key: string]: number
+}
+
+export interface Span extends Context {
+  trace_id: string
+  span_id: string
+  parent_id: string
+  name: string
+  resource: string
+  error: number
+  meta: SpanMetadata
+  metrics: SpanMetrics
+  start: number
+  duration: number
+  service: string
+  type: string
+}
+
+export interface TraceMetadata {
+  [key: string]: string
+}
+
+export interface Trace extends Context {
+  spans: Span[]
+  meta: TraceMetadata
+}
+
+interface Rum {
+  getInternalContext: (startTime?: number) => Context
+}
+
+export function startTraceCollection(userConfiguration: UserConfiguration, configuration: Configuration) {
+  if (!configuration.enableTracing) {
+    return
+  }
+  const batch = startTraceBatch(configuration)
+  const [, requestCompleteObservable] = startRequestCollection(configuration)
+
+  requestCompleteObservable.subscribe((data: RequestCompleteEvent) => {
+    if (data.traceId) {
+      batch.add(buildTrace(data, userConfiguration))
+    }
+  })
+}
+
+function startTraceBatch(configuration: Configuration) {
+  const primaryBatch = createTraceBatch(configuration.traceEndpoint)
+
+  let replicaBatch: Batch | undefined
+  if (configuration.replica !== undefined) {
+    replicaBatch = createTraceBatch(configuration.replica.traceEndpoint)
+  }
+
+  function createTraceBatch(endpointUrl: string) {
+    return new Batch(
+      new HttpRequest(endpointUrl, configuration.batchBytesLimit),
+      configuration.maxBatchSize,
+      configuration.batchBytesLimit,
+      configuration.maxMessageSize,
+      configuration.flushTimeout
+    )
+  }
+
+  return {
+    add(message: Trace) {
+      primaryBatch.add(message)
+      if (replicaBatch) {
+        replicaBatch.add(message)
+      }
+    },
+  }
+}
+
+function buildTrace(requestCompleteEvent: RequestCompleteEvent, userConfiguration: UserConfiguration) {
+  const traceId = requestCompleteEvent.traceId!.toString(16)
+  const globalMeta: SpanMetadata = {}
+
+  if (userConfiguration.env) {
+    globalMeta.env = userConfiguration.env
+  }
+  if (userConfiguration.service) {
+    globalMeta.service = `${userConfiguration.service}-http-client` // TODO what if not defined?
```

Comment:
@rochdev is service mandatory for tracing?

Reply:
Yes, I'm not 100% sure if the trace will be entirely dropped, but in any case most of APM won't work if the trace doesn't have a service.

Reply:
In `dd-trace` we fallback to `browser` if not defined.